File: Python/problem049.py
#! /usr/bin/env python3
'''
Problem 49 - Project Euler
http://projecteuler.net/index.php?section=problems&id=049
'''
from mathtools import primes
from itertools import combinations
from functools import reduce
from operator  import add

def samedigits(*xs):
    if len(xs) < 2:
        return True
    x = set(str(xs[0]))
    return all([set(str(y)) == x for y in xs[1:]])

def isarithmetical(*xs):
    if len(xs) < 2:
        return True
    xs = list(sorted(xs))
    ys = [xs[i+1]-xs[i] for i in range(len(xs)-1)]
    return all([z == ys[0] for z in ys[1:]])

if __name__ == '__main__':
    N = 9999    
    ps = list(filter(lambda x: x > 999, primes(N)))
    ans = [c for c in combinations(ps,3) if samedigits(*c) and isarithmetical(*c)]
    print(ans)
    print(list(map(lambda x: reduce(add, map(str, x)), ans)))


    

File: Go/problem049.go
package projecteuler

// #49 Prime permutations - Project Euler
// http://projecteuler.net/index.php?section=problems&id=49
import (
	"fmt"
	"projecteuler/my"
	"sort"
)

func P049() {
	ps := my.Primes(10000)

	m := map[int]struct{}{}
	for _, p := range ps {
		if p >= 1000 {
			m[p] = struct{}{}
		}
	}

	ans := make([]int, 3)
L:
	for _, p := range ps {
		if p < 1000 {
			continue
		}
		count := 0
		ans[0] = p

		a := make([]int, 4)
		for i := 3; i >= 0 && p > 0; i-- {
			a[i] = p % 10
			p /= 10
		}

		x := sort.IntSlice(a)
		for my.NextPermutation(x) {
			num := x[0]*1000 + x[1]*100 + x[2]*10 + x[3]
			if num == p {
				continue
			}

			if _, ok := m[num]; ok {
				if num == ans[count]+3330 {
					count += 1
					ans[count] = num
				}
			}

			if count == 2 && ans[0] != 1487 {
				break L
			}
		}
	}

	for _, x := range ans {
		fmt.Print(x)
	}
	fmt.Println("")
}
